server/channels/jobs/batch_migration_worker.go

```diff
@@ -64,7 +66,9 @@ func (worker *BatchMigrationWorker) Run() {
 	worker.logger.Debug("Worker started")
 	// We have to re-assign the stop channel again, because
 	// it might happen that the job was restarted due to a config change.
-	worker.stop = make(chan bool, 1)
+	if worker.closed.CompareAndSwap(true, false) {
```

Comment:
I'm wondering if we should put this at the top of the method and simply `return` if we fail to `CompareAndSwap`? The logic at present appears to allow `Run()` twice, duplicating the worker even though we guard `worker.stop`. I suspect there's also a cleaner way to approach all of this, but happy to defer so long as this is "correct".

Reply:
There are three cases to distinguish:
1. A freshly created worker
2. A `Run()`ing worker
3. A `Stop()`ed worker.

The code can't distinguish between 1 and 2 as both have `worker.closed` set to `false`. Moving the check to the top of the method doesn't help.

It's feasible to revert the semantics of `closed` to something like `running`, which means the default value `false` would apply for 1 and 3. But that seems like a large change that also needs to be applied to other similar code pieces. I'm hesitant to do that as part of this PR.

Reply:
@hanzei - Can you paste the race output here for us to understand what exactly was happening? This job worker is very tricky to get right and I'd love to have a unit test to lock in the behavior.

Reply:
@agnivade Sure:
```
WARNING: DATA RACE
Read at 0x00c006274038 by goroutine 252171:
  github.com/mattermost/mattermost/server/v8/channels/jobs.(*BatchMigrationWorker).Stop()
      /mattermost/server/channels/jobs/batch_migration_worker.go:88 +0x72
  github.com/mattermost/mattermost/server/v8/channels/jobs.(*Workers).Stop()
      /mattermost/server/channels/jobs/workers.go:84 +0x29b
  github.com/mattermost/mattermost/server/v8/channels/jobs.(*JobServer).StopWorkers()
      /mattermost/server/channels/jobs/server.go:111 +0x115
  github.com/mattermost/mattermost/server/v8/channels/app.(*Server).Shutdown()
      /mattermost/server/channels/app/server.go:756 +0x1867
  github.com/mattermost/mattermost/server/v8/channels/api4.(*TestHelper).ShutdownApp.func1()
      /mattermost/server/channels/api4/apitestlib.go:360 +0x52
{"timestamp":"2023-11-03 21:19:23.369 Z","level":"debug","msg":"Worker finished","caller":"jobs/base_workers.go:43","worker_name":"PostPersistentNotifications"}

Previous write at 0x00c006274038 by goroutine 252154:
  github.com/mattermost/mattermost/server/v8/channels/jobs.(*BatchMigrationWorker).Run()
      /mattermost/server/channels/jobs/batch_migration_worker.go:67 +0xb9
  github.com/mattermost/mattermost/server/v8/channels/jobs.(*Workers).Start.func1()
      /mattermost/server/channels/jobs/workers.go:52 +0x48

Goroutine 252171 (running) created at:
  github.com/mattermost/mattermost/server/v8/channels/api4.(*TestHelper).ShutdownApp()
      /mattermost/server/channels/api4/apitestlib.go:359 +0xfb
  github.com/mattermost/mattermost/server/v8/channels/api4.(*TestHelper).TearDown()
      /mattermost/server/channels/api4/apitestlib.go:378 +0xb1
  github.com/mattermost/mattermost/server/v8/channels/api4.TestUploadLicenseFile.func8()
      /mattermost/server/channels/api4/license_test.go:57 +0x39
  runtime.deferreturn()
      /usr/local/go/src/runtime/panic.go:476 +0x32
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
{"timestamp":"2023-11-03 21:19:23.369 Z","level":"debug","msg":"Worker stopping","caller":"jobs/base_workers.go:59","worker_name":"InstallNotifyAdmin"}
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47
```

Reply:
Oh I see, the `Stop` is being called from a different goroutine. Thanks.